### PR TITLE
Astro Adapter

### PR DESCRIPTION
**Pull request by tarekwiz**
I've created a functional adapter for .astro files which supports both the frontmatter and the JSX-like part of astro files.
This is based on the already present JSX adapter and has support for most of the edge cases that I could find.

### CODE REVIEW

**Comment by tarekwiz** [comment]
> Hi @tarekwiz, thank you for working on this!
> 
> I was trying to see how it works on a simple astro project and I was having issues. Please checkout https://github.com/wuchalejs/examples/tree/astro/astro and check why it complains. It assumes that you have this repo checkout next to the examples repo (for npm installs). I have set up the loading and setup already. And I want that dir to be the example for astro going forward once this gets merged.
> 
> Again, it's great that you're working on it!

There was an issue with nested components, which required major refactoring in the transformer and the runtime (changed from JSX to .astro which makes more sense), especially because Astro doesn't have {#snippet} like svelte does, so I had to create new components in the .wuchale folder to allow <slot> to work. I've also added some test cases. Please check and let me know if it works for you now.

**Comment by K1DV5** [comment]
The nested components work nicely now. It's much closer.

I added a language indicator message and there is one small issue, it seems expressions inside messages cause an error:
```astro
<p>You're viewing the {locale} page.</p>
```

Without the `{locale}` part, it works fine.

**Comment by tarekwiz** [comment]
I've pushed some changes to fix those cases, also found a few other edge cases and added tests for them. Let me know if you found any new edge case that needs fixing. I'm already running it in staging on a large website and it's doing great so far :). One thing I also need is some kind of way to extend the heuristic functions from outside the package so that one could for example exclude js object's key from the heuristic and only keep the value. This would be very useful for very specific cases. I'm still not sure what the best way to do this is, but I guess that's unrelated to the astro adapter.

File: packages/astro/src/transformer.ts [hunk 1]
```diff
@@ -0,0 +1,1526 @@
...
+  private static readonly SKIP_AST_KEYS = new Set(["type", "start", "end", "loc"]);
+
+  /** Cached parser instance */
+  private _parser: typeof Parser | null = null;
+  private _parserOpts: ReturnType<typeof scriptParseOptionsWithComments> | null = null;
+
+  /** Get TypeScript + JSX parser (lazy-initialized) */
+  private get parser() {
+    if (!this._parser) {
+      this._parser = Parser.extend(tsPlugin(), jsx());
```
Comment (K1DV5) [inline packages/astro/src/transformer.ts]: You don't actually need `acorn-jsx` for this. `tsPlugin` supports jsx like:
```js
      this._parser = Parser.extend(tsPlugin({jsx: true}))
```

File: packages/astro/src/transformer.ts [hunk 2]
```diff
@@ -0,0 +1,1526 @@
+import MagicString from "magic-string";
+import { Parser } from "acorn";
+import { Message } from "wuchale";
+import { tsPlugin } from "@sveltejs/acorn-typescript";
+import jsx from "acorn-jsx";
```
Comment (K1DV5) [inline packages/astro/src/transformer.ts]: So you can remove this. Causes a problem as it's not declared as a dep as well.

File: packages/astro/src/transformer.ts (hunk 1 above)
Comment (tarekwiz) [reply packages/astro/src/transformer.ts]: Nice catch.

File: packages/wuchale/src/adapters.ts [hunk 3]
```diff
@@ -188,9 +188,16 @@ export type TransformOutputFunc = (header: string) => {
     map?: any
 }
 
+export type AuxiliaryFile = {
```
Comment (K1DV5) [inline packages/wuchale/src/adapters.ts]: What is this?? Why would you need it?

Comment (K1DV5) [reply packages/wuchale/src/adapters.ts]: Once you make it self-contained (not touching the `wuchale` package or any other), we can merge it. Thanks!

Comment (K1DV5) [reply packages/wuchale/src/adapters.ts]: Ok I will look into this myself. It is okay if I modify the PR right? You have enabled it. I would like to work on this and the refactor before merging. I will make it in line with the other adapters.

Comment (K1DV5) [reply packages/wuchale/src/adapters.ts]: Also, since there is a pending release PR with other changes, I would like to base this on that in order to have the opportunity to change the base package if necessary and release it together. Agree?

Comment (tarekwiz) [reply packages/wuchale/src/adapters.ts]: Feel free to make any kind of changes :). I will push some changes I made to the test cases right now then you're free to do modifications.

Comment (K1DV5) [reply packages/wuchale/src/adapters.ts]: Thanks!

**Comment by K1DV5** [comment]
Hi @tarekwiz, I am working on this and gradually it became a rewrite as I needed to make it inline with the other adapters, and I finally found a way to not need auxiliary files! So I'm starting clean and including it in the next release branch #213. I will include you as a co-author in the new commits I make, because you did the original research for the information I'm using for the new implementation. I really appreciate your contribution!

**Comment by K1DV5** [comment]
Closing in favor of #213. Thanks again.

**Comment by K1DV5** [comment]
I just fixed some parsing errors, maybe you meant those. You can update `@wuchale/astro` and try it.

**Comment by tarekwiz** [comment]
I wrote some snippets which cause the error in my previous comment.

**Comment by K1DV5** [comment]
Fixed! you can update `@wuchale/astro` and try it. If there are further errors still, please let me know.

The problem was that in Golang (which astro's compiler is written in) the offsets are in bytes, and in js, they are in indices, which creates some differences in start and end ranges of nodes. Now there is a converter function to indices, so that everything is harmonious.
